### PR TITLE
Use 'gpconfig -u' to save unnecessary cluster restarting time.

### PR DESCRIPTION
**Pull request by higuoxing**
This patch teaches the demo cluster set plpython3.python_path to '$GPHOME/lib/python', so that we don't need to set it before running isolation2 tests. This saves some time for ICW tests.

Currently, Concourse is not happy with my change (I'll try to fix it).

https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-setup-python-path-for-demo-cluster-rocky8

### CODE REVIEW

**Review by soumyadeep2007 — APPROVED** [review]
LGTM!

**Comment by beeender** [comment]
Why this change is needed? regress/isolation2 tests should not use python libs from `$GPHOME/lib/python`. plpython and management-utilities may use different versions of python.

**Comment by higuoxing** [comment]
> Why this change is needed? regress/isolation2 tests should not use python libs from `$GPHOME/lib/python`. plpython and management-utilities may use different versions of python.

Makes sense. I cannot remember the motivation behind this PR ... I'll try to fix this PR after beta.5 release.

**Comment by ashwinstar** [comment]
> > Why this change is needed? regress/isolation2 tests should not use python libs from `$GPHOME/lib/python`. plpython and management-utilities may use different versions of python.
> 
> Makes sense. I cannot remember the motivation behind this PR ... I'll try to fix this PR after beta.5 release.

Isn't the reason that `src/test/isolation2/sql/setup.sql` has 

```
! gpconfig -c plpython3.python_path -v "'$GPHOME/lib/python'" --skipvalidation
! gpstop -ari;
```
which is adding unnecessary cluster restart time for setting this GUC hence the motivation for this PR

**Comment by higuoxing** [comment]
> > > Why this change is needed? regress/isolation2 tests should not use python libs from `$GPHOME/lib/python`. plpython and management-utilities may use different versions of python.
> > 
> > 
> > Makes sense. I cannot remember the motivation behind this PR ... I'll try to fix this PR after beta.5 release.
> 
> Isn't the reason that `src/test/isolation2/sql/setup.sql` has
> 
> ```
> ! gpconfig -c plpython3.python_path -v "'$GPHOME/lib/python'" --skipvalidation
> ! gpstop -ari;
> ```
> 
> which is adding unnecessary cluster restart time for setting this GUC hence the motivation for this PR

Haha, yes, it's the motivation for this PR. I mean I forget why we need to setup `python_python` for PL/Python scripts, since packages under the path `$GPHOME/lib/python` is for gpMgmt tools, right? Currently, gpMgmt tools (Python3.6) and PL/Python (Python3.9) are using different versions of interpreters, so it might not correct to set `python_path` to ` `$GPHOME/lib/python`. Please correct me if I make mistakes.

**Comment by higuoxing** [comment]
> > > > Why this change is needed? regress/isolation2 tests should not use python libs from `$GPHOME/lib/python`. plpython and management-utilities may use different versions of python.
> > > 
> > > 
> > > Makes sense. I cannot remember the motivation behind this PR ... I'll try to fix this PR after beta.5 release.
> > 
> > 
> > Isn't the reason that `src/test/isolation2/sql/setup.sql` has
> > ```
> > ! gpconfig -c plpython3.python_path -v "'$GPHOME/lib/python'" --skipvalidation
> > ! gpstop -ari;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which is adding unnecessary cluster restart time for setting this GUC hence the motivation for this PR
> 
> Haha, yes, it's the motivation for this PR. I mean I forget why we need to setup `python_python` for PL/Python scripts, since packages under the path `$GPHOME/lib/python` is for gpMgmt tools, right? Currently, gpMgmt tools (Python3.6) and PL/Python (Python3.9) are using different versions of interpreters, so it might not correct to set `python_path` to `$GPHOME/lib/python`. Please correct me if I make mistakes.

Some regression tests is using Python's subprocess to execute gpMgmt
tools. These tools require 'PYTHONPATH' being set. Hence, we need to
setup 'PYTHONPATH' when creating demo cluster. It might be worth
to refactor these tests since PL/Python shouldn't access
'$GPHOME/lib/python'.

**Comment by ashwinstar** [comment]
> 
> Some regression tests is using Python's subprocess to execute gpMgmt tools. These tools require 'PYTHONPATH' being set. Hence, we need to setup 'PYTHONPATH' when creating demo cluster. It might be worth to refactor these tests since PL/Python shouldn't access '$GPHOME/lib/python'.

Fun, Interesting. Please can you provide pointers to such tests. Also, what recommendation is proposed to refactor such tests.

Isn't a good practice though not sure if any customer workflows may also be coded on similar lines (like executing gpconfig to set some GUCs maybe from plpython UDF).

**Comment by higuoxing** [comment]
> Fun, Interesting. Please can you provide pointers to such tests.

Take this one for example,

https://github.com/greenplum-db/gpdb/blob/93435396e7dd0a3306b45039720a54e5ebe7accd/src/test/isolation2/sql/segwalrep/dtm_recovery_on_standby.sql#L88-L91

`gpinitstandby` is invoked in PL/Python's subprocess.

> Also, what recommendation is proposed to refactor such tests.

We have 2 optional solutions for now, neither of them is silver bullet.

1. `source greenplum_path.sh` before executing gpMgmt tools in PL/Python's subprocess.

   What we need to do is carefully checking each test case that invoking gpMgmt tools inside PL/Python and fixing them.

2. Make gpMgmt tools get rid of `PYTHONPATH`.

   Currently, we have removed all of the third-party Python packages from `PYTHONPATH`. The only one lives there is `gppylib`. Since `gppylib` is serving gpMgmt tools, my idea is moving `gppylib` to `$GPHOME/bin` and `$GPHOME/sbin` (where gpMgmt tools live), so that Python interpreter can automatically detect it. We have a draft patch for it (https://github.com/higuoxing/gpdb/tree/remove-python-path).

> Isn't a good practice though not sure if any customer workflows may also be coded on similar lines (like executing gpconfig to set some GUCs maybe from plpython UDF).

If we choose the second solution above, this issue can be addressed as well.

File: gpAux/gpdemo/demo_cluster.sh [hunk 1]
```diff
@@ -393,8 +393,10 @@ if [ "${BLDWRAP_POSTGRES_CONF_ADDONS}" != "__none__" ]  && \
     echo ${BLDWRAP_POSTGRES_CONF_ADDONS} | sed -e 's/\[//g' -e 's/\]//g' | tr "," "\n" | sed -e 's/^\"//g' -e 's/\"$//g' >> ${CLUSTER_CONFIG_POSTGRES_ADDONS}
 fi
 
-# Add fsync-off for all gpdemo deployments
+# Add fsync=off for all gpdemo deployments
 grep -q 'fsync=off' ${CLUSTER_CONFIG_POSTGRES_ADDONS} && echo "fsync=off already exists in ${CLUSTER_CONFIG_POSTGRES_ADDONS}." || echo "fsync=off" >> ${CLUSTER_CONFIG_POSTGRES_ADDONS}
+# Add plpython3.python_path='$GPHOME/lib/python' for all gpdemo deployments
```
Comment (beeender) [inline gpAux/gpdemo/demo_cluster.sh:398]: I still think it is much better to change the `plpython3.python_path` before tests start. The `gppylib` is only needed by some tests, not even all the tests. A normal user would never need it. And the demo cluster should be as same as possible with the GPDB release -- the GUC will be empty in the release. I think keeping the demo-cluster clean is easier for use to identify problem at an early stage.

Won't changing to `gpstop -u` enough to make this faster?

Comment (higuoxing) [reply gpAux/gpdemo/demo_cluster.sh:398]: Sounds reasonable.

**Review by beeender — APPROVED** [review]

**Comment by higuoxing** [comment]
I'm going to merge this PR.

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-setup-python-path-for-demo-cluster-rocky8

Failed tests is not caused by this PR.

**Pull request merged**